std/range/primitives.d

```diff
@@ -1664,6 +1674,154 @@ if (isInputRange!Range)
 }
 
 /**
+This is a best-effort implementation of `back` for any kind of
+range. It expects a non-empty range.
+
+If `isBidirectional!Range`, simply returns `range.back` without
+checking `upTo) (when specified).
+
+Otherwise, walks the range through its length and returns the last element seen.
+Performes $(BIGOH n) evaluations of `range.empty` and `range.popFront()`,
+where `n` is the effective length of `range`.
+
+The `upTo` parameter is useful to "cut the losses" in case
+the interest is in seeing whether the range has at least some number
+of elements. If the parameter `upTo` is specified, stops if `upTo`
+steps have been taken and returns `upTo`.
+
+Infinite ranges are compatible, provided the parameter `upTo` is
+specified, in which case the implementation simply returns `upTo`.
+
+Params:
+    range = range to walk through
+
+Returns: last element of the range
+
+See_Also: $(REF walkLength, std, range, primitives)
+*/
+auto walkBack(Range)(Range range)
+if ((isInputRange!Range || isIterable!Range) && !isInfinite!Range)
+in
+{
+
+    assert(!range.empty, "Can't walkBack an empty range");
```

Comment:
There are at least these options:
- use [Nullable](https://dlang.org/phobos/std_typecons.html#.Nullable)
- provide an (`.ifEmpty(value)`) method that will yield a range of length 1 with the default value if an empty range is passed in and the otherwise the range unmodified
- use `enforce` instead (generally disliked, because exceptions allocate)
- provide an overload with a `seed` 

The best precedent I found is [maxElement](https://dlang.org/phobos/std_algorithm_searching.html#.maxElement), so I guess we should follow this then.